sketch_oct9a.ino

```diff
@@ -119,160 +121,152 @@ void updateSDCardLEDs() {
+  last_voc = analogRead(6);
+
+  if (hum.getStatus() != HumiditySensor::ACTIVE) {
+    hum.initialize();
+    updateSensorLEDs();
   }
-  if (loc.longitude < 0) {
-    Serial.print(-loc.longitude);
-    Serial.print('W');
-  } else {
-    Serial.print(loc.longitude);
-    Serial.print('E');
+
+  float humid, temp;
+  if (hum.getValues(&humid, &temp)) {
+    last_humid = humid;
+    last_temp = temp;
   }
-  Serial.print(", ");
-  Serial.print(loc.altitudeMSL);
-  Serial.println("ft");
-
-  Serial.print(loc.timestamp.hours);
-  Serial.print(':');
-  if (loc.timestamp.minutes < 10) {
-    Serial.print('0');
+
+  // Requesting data from the humidity sensor can be relatively slow
+  yield();
+#endif
+
+  if (imu.getStatus() != IMU::ACTIVE) {
+    imu.initialize();
+    updateSensorLEDs();
   }
-  Serial.print(loc.timestamp.minutes);
-  Serial.print(':');
-  if (loc.timestamp.seconds < 10) {
-    Serial.print('0');
+
+  IMU::vector3 accel, gyro;
+  if (imu.getValues(&accel, &gyro)) {
+    last_accel = accel;
+    last_gyro = gyro;
   }
-  Serial.print(loc.timestamp.seconds);
-  Serial.print('.');
-  if (loc.timestamp.milliseconds < 100) {
-    Serial.print('0');
+
+  if (alt.getStatus() != Altimeter::ACTIVE) {
+    alt.initialize();
+    updateSensorLEDs();
   }
-  if (loc.timestamp.milliseconds < 10) {
-    Serial.print('0');
+
+  if (alt.getStatus() == Altimeter::ACTIVE) {
+    last_alt = alt.getAltitude();
   }
-  Serial.println(loc.timestamp.milliseconds);
-}
 
-void loop() {
-  delay(500);
+  yield();
 
+  // Commenting out SD code for now until it can be tested
+  /*
+  if (card.getStatus() != SDCard::ACTIVE) {
+    card.initialize();
+    updateSDCardLEDs();
+  }
+  card.writeToCSV(
+    last_coords,
+    last_accel,
+    last_alt,
 #if CAPSULE == 2
-  int v = analogRead(6);
-  Serial.print("VOC voltage: ");
-  Serial.println(3.3F / 1023.0F * (float)v);
+    last_voc,
+    last_humid,
+    last_temp,
 #endif
+    last_gyro
+  );
 
-  if (alt.getStatus() == Altimeter::ACTIVE) {
-    float altitude = alt.getAltitude();
-    Serial.print("Altitude (feet): ");
-    Serial.println(altitude);
-  } else {
-    alt.initialize();
-    updateSensorLEDs();
+  yield();
+  */
+}
+
+void gps_loop() {
+  if (gps.getStatus() != GPS::ACTIVE) {
+    gps.initialize();
+    updateGPSLEDs();
   }
 
-  if (imu.getStatus() == IMU::ACTIVE) {
-    IMU::vector3 accel;
-    IMU::vector3 gyro;
-    if (imu.getValues(&accel, &gyro)) {
-      Serial.print("Acceleration (m/s^2): ");
-      printAcceleration(accel);
-      Serial.print("\nGyroscope (rad/s): (");
-      Serial.print(gyro.x);
-      Serial.print(", ");
-      Serial.print(gyro.y);
-      Serial.print(", ");
-      Serial.print(gyro.z);
-      Serial.println(')');
-    } else {
-      Serial.println("No IMU value read");
-    }
-  } else {
-    imu.initialize();
-    updateSensorLEDs();
+  GPS::Coordinates c;
+  if (gps.getLocation(&c)) {
+    last_coords = c;
   }
 
+  const unsigned long GPS_FREQ = 18;
```

Comment:
Does this need defined every loop? Can this not be a const added at the top of the file instead of in this gps_loop?

Reply:
The optimizer will combine this with the line below it and just emit a single `delay(55);` statement. See [this example on Godbolt](https://godbolt.org/z/4zTGfa8ab).